src/delta.c

```diff
@@ -38,86 +38,164 @@
 {
-	char *origin;
-	char *dir, *base, *tmp = NULL, *tmp2 = NULL;
-	char *deltafile = NULL;
 	int ret;
-	struct stat sb;
 
-	string_or_die(&deltafile, "%s/delta/%i-%i-%s-%s", state_dir,
-		      file->deltapeer->last_change, file->last_change, file->deltapeer->hash, file->hash);
+	char *prefixed_from_file;
+	string_or_die(&prefixed_from_file, "%s/%s", path_prefix, from_file);
 
-	ret = stat(deltafile, &sb);
-	if (ret != 0) {
-		free_string(&deltafile);
-		return;
+	ret = apply_bsdiff_delta(prefixed_from_file, to_staged, delta_file);
+	if (ret) {
+		fprintf(stderr, "apply bsdiff delta failed\n");
+		goto fail;
 	}
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
-
-	dir = dirname(tmp);
-	base = basename(tmp2);
+	xattrs_copy(prefixed_from_file, to_staged);
 
-	string_or_die(&origin, "%s/%s/%s", path_prefix, dir, base);
-
-	ret = apply_bsdiff_delta(origin, filename, deltafile);
-	if (ret) {
-		unlink_all_staged_content(file);
-		goto out;
+	/* TODO: Remove this and rely only on hash verification only? */
+	if (xattrs_compare(prefixed_from_file, to_staged) != 0) {
+		fprintf(stderr, "xattrs comparison failed\n");
+		goto remove_and_fail;
 	}
-	xattrs_copy(origin, filename);
 
-	if (!verify_file(file, filename)) {
-		unlink_all_staged_content(file);
-		goto out;
+	char hash[SWUPD_HASH_LEN];
+	if (!compute_hash_from_file(to_staged, &hash[0])) {
+		fprintf(stderr, "hash calculation failed\n");
+		goto remove_and_fail;
 	}
+	if (!hash_equal(hash, to_hash)) {
+		fprintf(stderr, "hash mismatch\n");
+		goto remove_and_fail;
+	}
+
+	free_string(&prefixed_from_file);
+	return true;
+
+remove_and_fail:
+	(void)remove(to_staged);
+fail:
+	free_string(&prefixed_from_file);
 
-	if (xattrs_compare(origin, filename) != 0) {
-		unlink_all_staged_content(file);
-		goto out;
+	return false;
+}
+
+void apply_deltas(struct manifest *current_manifest)
+{
+	char *delta_dir;
+	string_or_die(&delta_dir, "%s/delta", state_dir);
+
+	DIR *dir = opendir(delta_dir);
+	if (!dir) {
+		/* No deltas available to apply. */
+		return;
 	}
 
-	unlink(deltafile);
+	struct dirent *ent;
+	while ((ent = readdir(dir))) {
+		if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, "..")) {
+			continue;
+		}
+
+		/* Delta files have the form VERSION-VERSION-FROM_HASH-TO_HASH, we need to
+		 * extract FROM and TO. */
+		char *delta_name = ent->d_name;
+		char *s = delta_name;
+
+		s = index(s, '-');
+		if (!s) {
+			fprintf(stderr, "Skipping invalid delta file %s/%s\n", delta_dir, delta_name);
+			continue;
+		}
+		s++;
+
+		s = index(s, '-');
+		if (!s) {
+			fprintf(stderr, "Skipping invalid delta file %s/%s\n", delta_dir, delta_name);
+			continue;
+		}
+		s++;
+
+		/* Note: SWUPD_HASH_LEN accounts for the nul-terminator after the hash. */
+
+		/* Two hashes (without zero) plus one dash in the middle. */
+		if (strlen(s) != ((2 * (SWUPD_HASH_LEN - 1)) + 1)) {
+			fprintf(stderr, "Skipping invalid delta file %s/%s\n", delta_dir, delta_name);
```

Comment:
Remove the invalid delta file as well?

Reply:
"Disaster prevention" flag: If we don't recognize the file, we don't remove it. Also let it around for further investigation (the deltas we know we can find in the server).

Reply:
What do you mean? It is an invalid delta file in the delta directory. afaict it won't get cleaned up and will just sit there. I have no idea why I would want to inspect that. This is the swupd state dir. If it can't do anything with the file the file should be removed. It similarly removes files with bad hashes in the staged dir.